### PR TITLE
chore: reject bad storm cache configurations

### PR DESCRIPTION
**Pull request by ajewellamz**
_Issue #, if available:_

_Description of changes:_

_Squash/merge commit message, if applicable:_

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/AwsCryptographyMaterialProvidersOperations.dfy [hunk 1]
```diff
@@ -270,6 +291,7 @@ module AwsCryptographyMaterialProvidersOperations refines AbstractAwsCryptograph
     var cmc;
 
     if input.cache.Some? {
+      var _ := CheckCache(input.cache.value, input.ttlSeconds);
```
Comment (seebees) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/AwsCryptographyMaterialProvidersOperations.dfy:294]: I think that all of these need to be `:-`

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/AwsCryptographyMaterialProvidersOperations.dfy [hunk 2]
```diff
@@ -258,6 +259,26 @@ module AwsCryptographyMaterialProvidersOperations refines AbstractAwsCryptograph
   predicate CreateAwsKmsHierarchicalKeyringEnsuresPublicly(input: CreateAwsKmsHierarchicalKeyringInput, output: Result<IKeyring, Error>)
   {true}
 
+  function method N(n : PositiveLong) : string {
+    String.Base10Int2String(n as int)
+  }
+
+  method CheckCache(cache : CacheType, ttlSeconds: PositiveLong) returns (output : Result<bool, Error>)
```
Comment (seebees) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/AwsCryptographyMaterialProvidersOperations.dfy:266]: Given that this function is returning a bool, I think that it is better modeled as an `Outcome`?

**Review by seebees — CHANGES_REQUESTED** [review]

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/AwsCryptographyMaterialProvidersOperations.dfy (hunk 2 above)
Comment (ajewellamz) [reply AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/AwsCryptographyMaterialProvidersOperations.dfy:266]: Whenever I use an Outcome, I eventually get bitten by the fact that Outcomes don't compose, and so I go back to Result. Now I just always start with Result.

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/AwsCryptographyMaterialProvidersOperations.dfy (hunk 1 above)
Comment (ajewellamz) [reply AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/AwsCryptographyMaterialProvidersOperations.dfy]: Whoops.

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/CMCs/StormTracker.dfy [hunk 3]
```diff
@@ -39,11 +40,65 @@ module {:options "/functionSyntax:4" }  StormTracker {
...
+  function CheckSettings(
+    cache: Types.StormTrackingCache
+  ) : (res : Result<bool, Types.Error> )
+    ensures res.Success? ==> ConsistentSettings(cache)
+  {
+    if ConsistentSettings(cache) then
+      Success(true)
+    else
+      Failure(Types.AwsCryptographicMaterialProvidersException(message := BadCacheMsg(cache)))
+  }
```
Comment (seebees) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/src/CMCs/StormTracker.dfy:99]: What about this?  I also feel like we could use a `Need` that runs the error message in a thunk, what do you think?

```suggestion
  function CheckSettings(
    cache: Types.StormTrackingCache
  ) : (res : Outcome<Types.Error> )
    ensures res.Pass? ==> ConsistentSettings(cache)
  {
    if ConsistentSettings(cache) then
      Pass
    else
      Fail(Types.AwsCryptographicMaterialProvidersException(message := BadCacheMsg(cache)))
  }
```

**Review by seebees — APPROVED** [review]
Thanks for adding the comments. LGTM.

The change in the default worried me, but I think it is ok.

**Pull request merged**